pkg/bootkube/create.go

```diff
@@ -43,143 +60,150 @@ func CreateAssets(config clientcmd.ClientConfig, manifestDir string, timeout tim
+		return nil, err
 	}
 
-	r := builder.
-		Schema(schema).
-		ContinueOnError().
-		NamespaceParam(cmdNamespace).DefaultNamespace().
-		FilenameParam(enforceNamespace, filenameOpts).
-		SelectorParam("").
-		Flatten().
-		Do()
-	err = r.Err()
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(c)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	count := 0
-	err = r.Visit(func(info *resource.Info, err error) error {
-		if err != nil {
-			return err
-		}
+	return &creater{
+		mapper: newResourceMapper(discoveryClient),
+		client: client,
+	}, nil
+}
 
-		obj, err := resource.NewHelper(info.Client, info.Mapping).Create(info.Namespace, true, info.Object)
-		if err != nil {
-			if apierrors.IsAlreadyExists(err) {
-				count++
-				return nil
-			}
-			return cmdutil.AddSourceToErr("creating", info.Source, err)
+func (c *creater) createManifests(manifests []manifest) (ok bool) {
+	// Bootkube used to create manifests in named order ("01-foo" before "02-foo").
+	// Maintain this behavior for everything except CRDs, which have strict ordering
+	// that we should always respect.
+	sort.Slice(manifests, func(i, j int) bool {
+		return manifests[i].filepath < manifests[j].filepath
+	})
+
+	var crds, other []manifest
+	for _, m := range manifests {
+		if m.kind == "CustomResourceDefinition" && strings.HasPrefix(m.apiVersion, "apiextensions.k8s.io/") {
+			crds = append(crds, m)
+		} else {
+			other = append(other, m)
 		}
-		info.Refresh(obj, true)
-		gvk := obj.GetObjectKind().GroupVersionKind()
+	}
 
-		// If the object is a CRD, wait for it to fully roll out before continuing.
-		// This is because we may also be creating other instances of this CRD later.
-		if gvk.Kind == "CustomResourceDefinition" {
-			if err = waitForCRDRollout(info.Client, gvk, info.Object); err != nil {
-				return err
-			}
+	create := func(m manifest) {
+		if err := c.create(m); err != nil {
+			ok = false
+			UserOutput("Failed creating %s: %v", m, err)
+		} else {
+			UserOutput("Created %s\n", m)
 		}
+	}
+
+	// Create the custom resource definition before creating the actual custom resources.
+	for _, m := range crds {
+		create(m)
+	}
 
-		count++
-		shortOutput := false
-		if !shortOutput {
-			f.PrintObjectSpecificMessage(info.Object, util.GlogWriter{})
+	// Wait until the API server registers the CRDs. Until then it's not safe to create the
+	// manifests for those custom resources.
+	for _, crd := range crds {
+		if err := c.waitForCRD(crd); err != nil {
+			ok = false
+			UserOutput("Failed waiting for %s: %v", crd, err)
 		}
-		cmdutil.PrintSuccess(mapper, shortOutput, util.GlogWriter{}, info.Mapping.Resource, info.Name, false, "created")
-		UserOutput("\tcreated %23s %s\n", info.Name, strings.TrimSuffix(info.Mapping.Resource, "s"))
-		return nil
-	})
-	if err != nil {
-		return err
 	}
-	if count == 0 {
-		return fmt.Errorf("no objects passed to create")
+
+	for _, m := range other {
+		create(m)
 	}
-	return nil
+	return true
```

Comment:
Woops, this needs to be `return ok`. Going to wait for CI to go green before fixing.